iroh/examples/transfer.rs

```diff
@@ -0,0 +1,313 @@
+    let cli = Cli::parse();
+
+    match &cli.command {
+        Commands::Provide { size, relay_url } => provide(*size, relay_url.clone()).await?,
+        Commands::Fetch { ticket, relay_url } => fetch(ticket, relay_url.clone()).await?,
+    }
+
+    Ok(())
+}
+
+async fn provide(size: u64, relay_url: Option<String>) -> anyhow::Result<()> {
+    let secret_key = SecretKey::generate();
+    let relay_mode = match relay_url {
+        Some(relay_url) => {
+            let relay_url = RelayUrl::from_str(&relay_url)?;
+            let relay_map = RelayMap::from_url(relay_url);
+            RelayMode::Custom(relay_map)
+        }
+        None => RelayMode::Default,
+    };
+    let endpoint = Endpoint::builder()
+        .secret_key(secret_key)
+        .alpns(vec![TRANSFER_ALPN.to_vec()])
+        .relay_mode(relay_mode)
+        .bind()
+        .await?;
+
+    let node_id = endpoint.node_id();
+
+    for local_endpoint in endpoint
+        .direct_addresses()
+        .next()
+        .await
+        .context("no endpoints")?
+    {
+        println!("\t{}", local_endpoint.addr)
+    }
+
+    let relay_url = endpoint
+        .home_relay()
+        .expect("should be connected to a relay server");
+    let local_addrs = endpoint
+        .direct_addresses()
+        .next()
+        .await
+        .context("no endpoints")?
+        .into_iter()
+        .map(|endpoint| endpoint.addr)
+        .collect::<Vec<_>>();
+
+    let node_addr = NodeAddr::from_parts(node_id, Some(relay_url), local_addrs);
+    let ticket = NodeTicket::new(node_addr);
+
+    println!("NodeTicket: {}", ticket);
+
+    // accept incoming connections, returns a normal QUIC connection
+    while let Some(incoming) = endpoint.accept().await {
+        let mut connecting = match incoming.accept() {
+            Ok(connecting) => connecting,
+            Err(err) => {
+                tracing::warn!("incoming connection failed: {err:#}");
+                // we can carry on in these cases:
+                // this can be caused by retransmitted datagrams
+                continue;
+            }
+        };
+        let alpn = connecting.alpn().await?;
+        let conn = connecting.await?;
+        let node_id = iroh_net::endpoint::get_remote_node_id(&conn)?;
+        info!(
+            "new connection from {node_id} with ALPN {} (coming from {})",
+            String::from_utf8_lossy(&alpn),
+            conn.remote_address()
+        );
+
+        // spawn a task to handle reading and writing off of the connection
+        tokio::spawn(async move {
+            // accept a bi-directional QUIC connection
+            // use the `quinn` APIs to send and recv content
+            let (mut send, mut recv) = conn.accept_bi().await?;
+            tracing::debug!("accepted bi stream, waiting for data...");
+            let message = recv.read_to_end(100).await?;
+            let message = String::from_utf8(message)?;
+            println!("received: {message}");
```

Comment:
This seems expensive for something that's meant to measure the transfer speed.  Printing to terminals (or whatever output) is notoriously slow.  Checking valid UTF-8 is also slow.

Reply:
The example itself is slightly verbose so that you know things are moving around when other folks play with it. It does not print any of the simulated data transfer. It's a cost Im willing to accept for a better user experience.

Reply:
Yeah, fair.  Given the amount of data this doesn't actually matter as much, I didn't really realise this yet when I wrote this.